doc/designs/expired_certificate_pruning.md

```diff
@@ -0,0 +1,282 @@
+
+### Commands
+
+This will be implemented in the ipa-acme-manage command. While strictly not completely ACME-related this is the primary driver for pruning.
+
+A new verb will be added, pruning, to be used for enabling and configuring pruning.
+
+### Enabling pruning
+
+`$ ipa-acme-manage pruning --enable=TRUE`
+
+Enabling the job will call
+
+`$ pki-server ca-config-set jobsScheduler.job.pruning.enabled true`
+
+This will also set jobsScheduler.job.pruning.cron to `"0 0 1 * *"` if it has not already been set.
+
+Additionally it will set the job owner to `ipara` with:
+
+`pki-server ca-config-set jobsScheduler.job.pruning.owner ipara`
+
+Disabling the job will call
+
+`$ pki-server ca-config-unset jobsScheduler.job.pruning.enabled`
+
+### Cron settings
+
+To modify the cron settings:
+
+`$ ipa-acme-manage pruning --cron="Minute Hour Day_of_month Month_of_year Day_of_week"`
+
+Validation of the value will be:
+* each of the options is an integer
+* minute is within 0-59
+* hour is within 0-23
+* day of month is within 0-31
+* month of year is within 1-12
+* day of week is within 0-6
+
+No validation of setting February 31st will be done. That will be left to PKI. Buyer beware.
+
+### Disabling pruning
+
+`$ ipa-acme-manage pruning --enable=FALSE`
+
+This will remove the configuration option for `jobsScheduler.job.pruning.cron` just to be sure it no longer runs.
+
+### Configuration
+
+#### Pruning certificates
+
+`$ ipa-acme-manage pruning --certretention=VALUE --certretentionunit=UNIT`
+
+will be the equivalent of:
+
+`$ pki-server ca-config-set jobsScheduler.job.pruning.certRetentionTime 30`
+
+`$ pki-server ca-config-set jobsScheduler.job.pruning.certRetentionUnit day`
+
+The unit will always be required when modifying the time.
+
+`$ ipa-acme-manage pruning --certsearchsizelimit=VALUE --certsearchtimelimit=VALUE`
+
+will be the equivalent of:
+
+`$ pki-server ca-config-set jobsScheduler.job.pruning.certSearchSizeLimit 1000`
+
+`$ pki-server ca-config-set jobsScheduler.job.pruning.certSearchTimeLimit 0`
+
+A value of 0 for searchtimelimit is unlimited.
+
+#### Pruning requests
+
+`$ ipa-acme-manage pruning --requestretention=VALUE --requestretentionunit=UNIT`
+
+will be the equivalent of:
+
+`$ pki-server ca-config-set jobsScheduler.job.pruning.requestRetentionTime 30`
+
+`$ pki-server ca-config-set jobsScheduler.job.pruning.requestRetentionUnit day`
+
+The unit will always be required when modifying the time.
+
+`$ ipa-acme-manage pruning --requestsearchsizelimit=VALUE --requestsearchtimelimit=VALUE`
+
+
+will be the equivalent of:
+
+`$ pki-server ca-config-set jobsScheduler.job.pruning.requestSearchSizeLimit 1000`
+
+`$ pki-server ca-config-set jobsScheduler.job.pruning.requestSearchTimeLimit 0`
+
+A value of 0 for searchtimelimit is unlimited.
```

Comment:
I assume that the searchtimelimit and searchsizelimit are client-side values but the results can also be limited by server-side limits. Which user is ultimately performing the search on the LDAP server? Does pki use uid=pkidbuser? If that's the case it would be worth mentioning that the limits provided through `ipa-acme-manage pruning --requestsearchsizelimit=VALUE --requestsearchtimelimit=VALUE` cannot override the limits for this user.

Reply:
pki binds with internaldb.ldapauth.bindDN in CS.cfg which is Directory Manager, so no limits.

Reply:
On my instance, I have by default:
```
internaldb.ldapauth.authtype=SslClientAuth
internaldb.ldapauth.bindDN=cn=Directory Manager
internaldb.ldapauth.bindPWPrompt=internaldb
internaldb.ldapauth.clientCertNickname=subsystemCert cert-pki-ca
internaldb.ldapconn.host=server.ipa.test
internaldb.ldapconn.port=636
internaldb.ldapconn.secureConn=true
```
and all the connections between pki and the LDAP server are authenticated with `subsystemCert cert-pki-ca`, linked to uid=pkidbuser,ou=people,o=ipaca.

Reply:
internaldb.ldapauth.bindDN is what @edewata told me. It may be because its an internal task.

Reply:
I think it depends on the `internaldb.ldapauth.authtype`. If it's set to `BasicAuth` (which is the default) it will use `internaldb.ldapauth.bindDN`. If it's set to `SslClientAuth` it will use the user mapped to the cert specified in `internaldb.ldapauth.clientCertNickname`.

Reply:
So I don't think the `internaldb.ldapauth.bindDN` and `internaldb.ldapauth.bindPWPrompt` are needed in IPA case.

Reply:
Thanks. I added a mention of the limits and a pointer to the 389-ds tuning documentation.